File: 235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree/lowestCommonAncestor.py
"""
For a root, 
if (root.val - p.val) * (root.val - q.val) < 0, root must be their common ancestor.
if (root.val - p.val) * (root.val - q.val) > 0, p and q belong to the same subtree. 
if (root.val - p.val) * (root.val - q.val) = 0, root must be the common ancestor.
"""
class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        while (root.val - p.val) * (root.val - q.val) > 0:
            root = (root.left, root.right)[p.val > root.val]
        return root

File: 235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree/lowestCommonAncestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// O(lg(n))
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        while((p->val - root->val) * (q->val - root->val) > 0) {
            if(p->val > root-> val)
                root = root->right;
            else
                root = root->left;
        }
        return root;
    }
};
